include/sycl/algorithm/reduce.hpp

```diff
@@ -94,6 +100,343 @@ typename std::iterator_traits<Iterator>::value_type reduce(
+  size_t local_mem_size =
+    device.get_info<cl::sycl::info::device::local_mem_size>();
+  //std::cout << "local_mem_size=\t" << local_mem_size << std::endl;
+  size_t nb_work_item = min(max_work_item, local_mem_size / sizeofB);
+  //std::cout << "nb_work_item=\t" << nb_work_item << std::endl;
+
+  /* (nb_work_item == 0) iff (sizeof(T) > local_mem_size)
+   * If sizeof(T) > local_mem_size, this means that an object
+   * of type T can't hold in the memory of a single work-group
+   */
+  if (nb_work_item == 0) {
+    return mapreduce_descriptor(size, 0, 0, 0);
+  }
+  // we ensure that each work_item of every work_group is used at least once
+  size_t nb_work_group = min(max_work_group,
+                             up_rounded_division(size, nb_work_item));
+  //std::cout << "nb_work_group=\t" << nb_work_group << std::endl;
+  assert(nb_work_group >= 1);
+
+  //number of elements manipulated by each work_item
+  size_t size_per_work_item =
+    up_rounded_division(size, nb_work_item * nb_work_group);
+  //std::cout << "size_per_work_item=\t" << size_per_work_item << std::endl;
+  //number of elements manipulated by each work_group (except the last one)
+  size_t size_per_work_group = size_per_work_item * nb_work_item;
+  //std::cout << "size_per_work_group=\t" << size_per_work_group << std::endl;
+
+  nb_work_group = max(static_cast<size_t>(1),
+                      up_rounded_division(size, size_per_work_group));
+  //std::cout << "nb_work_group=\t" << nb_work_group
+  //          << " (updated)" << std::endl;
+  assert(nb_work_group >= 1);
+
+  assert(size_per_work_group * (nb_work_group - 1) < size);
+  assert(size_per_work_group * nb_work_group >= size);
+  /* number of elements manipulated by the last work_group
+   * n.b. if the value is 0, the last work_group is regular
+   */
+  size_t size_last_work_group = size % size_per_work_group;
+  //std::cout << "size_last_work_group=" << size_last_work_group << std::endl;
+
+  size_t size_per_work_item_last = up_rounded_division(size_last_work_group,
+                                                       nb_work_item);
+
+  return mapreduce_descriptor(size, size_per_work_group, nb_work_group,
+                              nb_work_item);
+}
+
+/*
+ * MapReduce Algorithm applied on a buffer
+ */
+
+template <typename ExecutionPolicy,
+          typename A,
+          typename B,
+          typename Reduce,
+          typename Map>
+B buffer_mapreduce(ExecutionPolicy &snp,
+                   cl::sycl::queue q,
+                   cl::sycl::buffer<A, 1> input_buff,
+                   B init, //map is not applied on init
+                   mapreduce_descriptor d,
+                   Map map,
+                   Reduce reduce) {
+
+  size_t size = d.size;
+  size_t size_per_work_group = d.size_per_work_group;
+  size_t nb_work_group = d.nb_work_group;
+  size_t nb_work_item = d.nb_work_item;
+  /*
+   * 'map' is not applied on init
```

Comment:
I have the feeling this comment (with a lot of *. Is it the common style in the PSTL?) would benefit to the full function instead.